Explanation of custom queries (#170)
mramendi/kie-docs: doc-content/shared-kie-docs/src/main/asciidoc/KieServer/kie-server-rest-api-custom-query-endpoints-ref.adoc

[id='kie-server-rest-api-custom-query-endpoints-ref_{context}']
= Custom query endpoints

You can use custom query endpoints to create and access custom queries in {PRODUCT}. Custom queries can request any data from the {PRODUCT} database.

> A number of custom queries are included in {PRODUCT}. You can use these queries to access complete lists of process instances, human tasks, and so on.

Reviewer: ```suggestion
A number of custom queries are included in {PRODUCT}. You can use these queries to access complete lists of process instances and user tasks.
```

> You can use the `/server/queries/definitions` GET endpoint to view the list of all custom queries. Use `/server/queries/definitions/{QueryName}` PUT and POST endpoints to define new custom queries.

Reviewer: I think this information is covered in the Swagger source.

> To run a custom query and return its results, use the `/server/queries/definitions/{queryName}/data`, `/server/queries/definitions/{queryName}/filtered-data`, and `/server/queries/definitions/containers/{containerId}/query/{queryName}/filtered-data` endpoints.

Reviewer: This one is also covered in the Swagger source.

When running a custom query, you must provide the name of a _query mapper_ in the `mapper` parameter. A mapper maps SQL query results into objects for a JSON response. You can implement your own query result mappers or use the mappers provided with {PRODUCT}. The query mappers in {PRODUCT} are similar to other object-relational mapping (ORM) providers, such as Hibernate, which maps tables to entities.

> For example, if a custom query returns process instance data, you can use the `org.jbpm.kie.services.impl.query.mapper.ProcessInstanceQueryMapper` mapper, also registered as `ProcessInstances`. if a custom query returns human task data, you can use the `org.jbpm.kie.services.impl.query.mapper.UserTaskInstanceQueryMapper` mapper, also registered as `UserTasks`. Youc an also use other mappers that provide additional information.

Reviewer: ```suggestion
For example, if a custom query returns process instance data, you can use the `org.jbpm.kie.services.impl.query.mapper.ProcessInstanceQueryMapper` mapper, also registered as `ProcessInstances`. if a custom query returns human task data, you can use the `org.jbpm.kie.services.impl.query.mapper.UserTaskInstanceQueryMapper` mapper, also registered as `UserTasks`. You can also use other mappers that provide additional information.
```
